redwood-flexbox-engine/src/commonMain/kotlin/app/cash/redwood/FlexboxEngine.kt

```diff
@@ -0,0 +1,2041 @@
+    return if (isMainHorizontal) node.margin.end else node.margin.bottom
+  }
+
+  /**
+   * Returns the flexItem's start margin in the cross axis. Either start or top.
+   *
+   * @param node the flexItem
+   * @param isMainHorizontal is the main axis horizontal
+   * @return the flexItem's start margin in the cross axis
+   */
+  private fun getFlexItemMarginStartCross(node: Node, isMainHorizontal: Boolean): Int {
+    return if (isMainHorizontal) node.margin.top else node.margin.start
+  }
+
+  /**
+   * Returns the flexItem's end margin in the cross axis. Either end or bottom.
+   *
+   * @param node the flexItem
+   * @param isMainHorizontal is the main axis horizontal
+   * @return the flexItem's end margin in the cross axis
+   */
+  private fun getFlexItemMarginEndCross(node: Node, isMainHorizontal: Boolean): Int {
+    return if (isMainHorizontal) node.margin.bottom else node.margin.end
+  }
+
+  /**
+   * Determine if a wrap is required (add a new flex line).
+   *
+   * @param mode the width or height mode along the main axis direction
+   * @param maxSize the max size along the main axis direction
+   * @param currentLength the accumulated current length
+   * @param childLength the length of a child view which is to be collected to the flex line
+   * @param flexItem the LayoutParams for the view being determined whether a new flex line is needed
+   * @param flexLinesSize the number of the existing flexlines size
+   * @return `true` if a wrap is required, `false` otherwise
+   */
+  private fun isWrapRequired(
+    mode: MeasureSpecMode,
+    maxSize: Int,
+    currentLength: Int,
+    childLength: Int,
+    flexItem: Node,
+    flexLinesSize: Int,
+  ): Boolean {
+    if (flexWrap == FlexWrap.NoWrap) {
+      return false
+    }
+    if (flexItem.wrapBefore) {
+      return true
+    }
+    if (mode == MeasureSpecMode.Unspecified) {
+      return false
+    }
+    // Judge the condition by adding 1 to the current flexLinesSize because the flex line
+    // being computed isn't added to the flexLinesSize.
+    if (maxLines != -1 && maxLines <= flexLinesSize + 1) {
+      return false
+    }
+    return maxSize < currentLength + childLength
+  }
+
+  private fun isLastFlexItem(
+    childIndex: Int,
+    childCount: Int,
+    flexLine: Line,
+  ) = childIndex == childCount - 1 && flexLine.itemCountVisible > 0
+
+  private fun addFlexLine(
+    flexLines: MutableList<Line>,
+    flexLine: Line,
+    viewIndex: Int,
+    usedCrossSizeSoFar: Int,
+  ) {
+    flexLine.sumCrossSizeBefore = usedCrossSizeSoFar
+    flexLine.lastIndex = viewIndex
+    flexLines.add(flexLine)
+  }
+
+  /**
+   * Checks if the view's width/height don't violate the minimum/maximum size constraints imposed
+   * by the [Node.minWidth], [Node.minHeight], [Node.maxWidth] and [Node.maxHeight] attributes.
+   *
+   * @param node the view to be checked
+   * @param index index of the view
+   */
+  private fun checkSizeConstraints(node: Node, index: Int) {
```

Comment:
I hate this function. In particular, I hate that its name is prefixed `check()` but it neither returns a boolean nor does it throw an exception if the check fails. I think this needs a new name; maybe `measureIfNecessary()` ?